numpy/core/src/umath/ufunc_object.c

```diff
@@ -991,33 +992,45 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
  */
 static int
 check_for_trivial_loop(PyUFuncObject *ufunc,
-                        PyArrayObject **op,
-                        PyArray_Descr **dtype,
-                        npy_intp buffersize)
+        PyArrayObject **op, PyArray_Descr **dtypes,
+        npy_intp buffersize)
 {
-    npy_intp i, nin = ufunc->nin, nop = nin + ufunc->nout;
+    int i, nin = ufunc->nin, nop = nin + ufunc->nout;
 
     for (i = 0; i < nop; ++i) {
         /*
          * If the dtype doesn't match, or the array isn't aligned,
          * indicate that the trivial loop can't be done.
          */
-        if (op[i] != NULL &&
-                (!PyArray_ISALIGNED(op[i]) ||
-                !PyArray_EquivTypes(dtype[i], PyArray_DESCR(op[i]))
-                                        )) {
+        if (op[i] == NULL) {
+            continue;
+        }
+        int must_copy = !PyArray_ISALIGNED(op[i]);
+
+        if (dtypes[i] != PyArray_DESCR(op[i])) {
+            NPY_CASTING safety = PyArray_GetCastSafety(
+                    PyArray_DESCR(op[i]), dtypes[i], NULL);
+            if (safety < 0) {
+                /* A proper error during a cast check should be rare */
+                return -1;
+            }
+            if (!(safety & _NPY_CAST_IS_VIEW)) {
+                must_copy = 1;
+            }
+        }
+        if (must_copy) {
```

Comment:
It is actually a bit weird that the copy is done here, possibly making a copy when the next argument could fail to be trivial and thus the effort wasted. Though I guess the logic is that the cast has to be done anyway even for a non-trivial loop and it might as well be done here? It still seems surprising for a `check` function...

Anyway, that was here already, so no big deal!

Reply:
The error paths are going to be oddly slow though. Otherwise its odd, but the effort is at least not quite wasted since it will safe work in `NpyIter`.

Reply:
I guess a slow error is not bad, and as you note, the effort is not wasted if everything succeeds. It remains odd to do it in a `check_*` function, but out of scope here!